crates/bevy_pbr/src/render/mesh.rs

```diff
@@ -384,60 +389,183 @@ pub fn extract_skinned_meshes(
     commands.insert_or_spawn_batch(values);
 }
 
+/// Data necessary to be equal for two draw commands to be mergeable
+///
+/// This is based on the following assumptions:
+/// - Only entities with prepared assets (pipelines, materials, meshes) are
+///   queued to phases
+/// - View bindings are constant across a phase for a given draw function as
+///   phases are per-view
+/// - `prepare_mesh_uniforms` is the only system that performs this batching
+///   and has sole responsibility for preparing the per-object data. As such
+///   the mesh binding and dynamic offsets are assumed to only be variable as a
+///   result of the `prepare_mesh_uniforms` system, e.g. due to having to split
+///   data across separate uniform bindings within the same buffer due to the
+///   maximum uniform buffer binding size.
+#[derive(Default, PartialEq, Eq)]
+struct BatchMeta<'mat, 'mesh> {
+    /// The pipeline id encompasses all pipeline configuration including vertex
+    /// buffers and layouts, shaders and their specializations, bind group
+    /// layouts, etc.
+    pipeline_id: Option<CachedRenderPipelineId>,
+    /// The draw function id defines the RenderCommands that are called to
+    /// set the pipeline and bindings, and make the draw command
+    draw_function_id: Option<DrawFunctionId>,
+    /// The material binding meta includes the material bind group id and
+    /// dynamic offsets.
+    material_binding_meta: Option<&'mat MaterialBindGroupId>,
+    mesh_handle: Option<&'mesh Handle<Mesh>>,
+    mesh_flags: u32,
+    dynamic_offset: Option<NonMaxU32>,
+}
+
+impl<'mat, 'mesh> BatchMeta<'mat, 'mesh> {
+    #[inline]
+    fn matches(&self, other: &BatchMeta<'mat, 'mesh>, consider_material: bool) -> bool {
+        self.pipeline_id == other.pipeline_id
+            && self.draw_function_id == other.draw_function_id
+            && self.mesh_handle == other.mesh_handle
+            && (self.mesh_flags & (MeshFlags::SKINNED | MeshFlags::MORPH_TARGETS).bits()) == 0
```

Comment:
The problem is rather that `many_foxes` draws many skinned meshes in sequence. These will have the same pipeline id, but critically they need to have separate dynamic offset bindings for the skinning data. So this is basically a way to make sure that skinned mesh draws that are otherwise batchable, do not get batched, because we know that the draws have to be separate to either update bindings or dynamic offsets. If PhaseItem _were_ more like a draw struct and contained bind group ids and dynamic offsets, then this would have been caught automatically.

Reply:
ahh got it, i misread it as batching skinned items together and unskinned items together, but it is actually not batching skinned/morphed items at all. 

so if we didn't use dynamic offsets, but bound the full buffer and used the instance id to look up into it then it would work? at least for skinning. we would need the morph targets textures to be the same size.

then we'd probably want to relax the 256 max joints and make the skinning array size based on sum of joints in the batch, and pass an index and count in the instance data, and probably move morph weights into a single batch array with a similar treatment as well, and then maybe consider whether we actually prefer to batch non-skinned and skinned together, and it's definitely a separate pr...

Reply:
The 256 comes from 16kB / 64 bytes per Mat4. So it’s the maximum uniform buffer binding size that’s limiting here. However, we could use an array in a data texture and index into it that way.